mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -163,25 +190,28 @@ class StringIndexerModel (
     }
     transformSchema(dataset.schema, logging = true)
 
+    val metadata = NominalAttribute.defaultAttr
+      .withName($(outputCol)).withValues(labels).toMetadata()
```

Comment:
withValues should include a special field for invalid, if we are keeping invalid labels.  How about calling that field "_invalidLabels" (and defining this constant in the StringIndexer object)?